bazel/cc_toolchains/clang_configuration.bzl

```diff
@@ -26,18 +26,23 @@ def _detect_system_clang(repository_ctx):
     # If the user provides an explicit `CC` environment variable, use that as
     # the compiler. This should point at the `clang` executable to use.
     cc = repository_ctx.os.environ.get("CC")
+    cc_path = None
     if cc:
-        version_output = _run(repository_ctx, [cc, "--version"]).stdout
-        if not "clang" in version_output:
-            fail("The `CC` environment variable is not a Clang compiler.")
-        return repository_ctx.path(cc)
-
-    # Try looking on the path. We only check for the normal name.
-    system_clang = repository_ctx.which("clang")
-    if not system_clang:
+        cc_path = repository_ctx.path(cc)
+        if not cc_path.exists:
+            cc_path = repository_ctx.which(cc)
+    if not (cc_path and cc_path.exists):
+        cc_path = repository_ctx.which("clang")
+    if not (cc_path and cc_path.exists):
         fail("Unable to find a `clang` executable on the system path.")
-
-    return system_clang
+    if cc_path and cc_path.exists:
+        version_output = _run(repository_ctx, [cc_path, "--version"]).stdout
+        if "clang" not in version_output:
+            if cc:
+                fail("The `CC` environment variable `%s` resolved to `%s` which is not a Clang compiler." % cc_path)
+            else:
+                fail("Tye `clang` executable on the system path resolved to `%s` which does not appear to be a Clang compiler" % cc_path)
```

Comment:
I'm concerned this approach is starting to separate the logic paths and resulting error a little too much. In particular, consider the case where:

1) CC is set
2) `clang` is in the PATH
3) `clang` is something not the compiler

I believe this would use the `if cc` error path, but provide the PATH resolution as the detected path, which it shouldn't. I'd lean towards going back to something closer to the previous approach. Doing something like adding `_is_clang` helper function to encapsulate the `--version` check may help, but I think there may be limits to how much duplication should be avoided here.